html/user/recover_email.php

```diff
@@ -0,0 +1,64 @@
+<?php
+// This file is part of BOINC.
+// http://boinc.berkeley.edu
+// Copyright (C) 2008 University of California
+//
+// BOINC is free software; you can redistribute it and/or modify it
+// under the terms of the GNU Lesser General Public License
+// as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// BOINC is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+// See the GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+
+require_once("../inc/util.inc");
+require_once("../inc/token.inc");
+require_once("../inc/email.inc");
+require_once("../inc/password_compat/password.inc");
+
+check_get_args(array("id", "token"));
+
+redirect_to_secure_url("recover_email.php");
+
+$user = get_logged_in_user(false);
+//Log out to clear all auth tokens
+if ($user) {
+    clear_cookie('auth');
+    $g_logged_in_user = null;
+}
+
+page_head(tra("Recover email address"));
+
+$userid = get_int("id", true);
+$token = get_str("token", true);
+
+if(is_valid_token($userid, $token, TOKEN_TYPE_CHANGE_EMAIL)) {
+    $tmpuser = BoincUser::lookup_id_nocache($userid);
+    //We can only change passwd_hash if we can get the userdata. 
+    if($tmpuser) {
```

Comment:
You might to want add an `else` just in case.

Reply:
I'm going to put in this invalid token as well.  Because letting someone know a userid is invalid may give them too much information.